pkg/storage/util.go

```diff
@@ -106,23 +104,19 @@ func NamespaceKeyFunc(prefix string, obj runtime.Object) (string, error) {
 	if err != nil {
 		return "", err
 	}
-	name := meta.GetName()
-	if msgs := validation.IsValidPathSegmentName(name); len(msgs) != 0 {
-		return "", fmt.Errorf("invalid name: %v", msgs)
-	}
-	return prefix + "/" + meta.GetNamespace() + "/" + name, nil
+	// We assume that NamespaceKeyFunc is used for read requests (List and Watch)
+	// and thus the object is correct and we don't perform any validation.
+	return prefix + "/" + meta.GetNamespace() + "/" + meta.GetName(), nil
 }
 
 func NoNamespaceKeyFunc(prefix string, obj runtime.Object) (string, error) {
 	meta, err := meta.Accessor(obj)
 	if err != nil {
 		return "", err
 	}
-	name := meta.GetName()
-	if msgs := validation.IsValidPathSegmentName(name); len(msgs) != 0 {
```

Comment:
>
it looks like that helps watches that filter on label/field selectors, but not ones that filter on key prefix, so most watches (that don't specify a labelSelector or fieldSelector) would still have to filter all watch events from all namespaces, and they each currently invoke keyFunc() on each watch event.

@liggitt - this isn't 100% true. Basically for the prefix-matching filter we have only two cases:
- all objects from a given namespace
- only one object
For the former yes - my optimization doesn't help. For the latter it actually helps, because this is expressed as a filter selector.